Automattic/wp-calypso
# Masterbar Site Menu - show whenever a site is selected and enable having selected site at the /sites dashboard.

Related to # https://github.com/Automattic/dotcom-forge/issues/8298 and https://github.com/Automattic/dotcom-forge/issues/8234

## Proposed Changes

* Removes the check for `sectionGroup === 'sites'` to restrict showing the site menu. Instead we show it any time a site is selected.
* Removes the check for `sectionGroup === 'sites'` to have the +new (site actions) use URLs for the currently selected site. Instead we will always use the currently selected site, and fallback to the site selector only when no site is selected.
* Removes the dashboards controller call to `setAllSitesSelected()` on initial load. This is no longer necessary and was being overridden by the below hook anyways.
* Updates the `use-sync-selected-site` hook to not clear the selected site on initial load of the dashboard. 
* Updates the `getShouldShowCollapsedGlobalSidebar` selector to not be based on immediate selected siteID value. 
    * Previously, when we were on /sites and had a site selected this selected assumed it should start showing the collapsed sidebar instead, for the purposes of triggering the collapse animation early enough to not be jumpy. 
    * After our change, this caused the collapsed sidebar to always show up in the dashboard when having a site selected. I have refactored this to use a local object state to determine if we have just selected a site, so we do not show the collapsed sidebar at the time and continue to have a good animation.

BEFORE
![image](https://github.com/user-attachments/assets/aced19f0-f57d-4665-a7b5-d823cc9485ae)

AFTER
<img width="679" alt="Screenshot 2024-07-16 at 10 34 53 AM" src="https://github.com/user-attachments/assets/d3b90865-a9fb-4d31-9a71-f6aff85a6342">

AFTER (my home to reader)
<img width="594" alt="Screenshot 2024-07-16 at 10 35 29 AM" src="https://github.com/user-attachments/assets/21854b5a-fd76-491f-9bb7-c95a2a0db6fa">

AFTER (my home to /sites)
<img width="1116" alt="Screenshot 2024-07-16 at 1 05 47 PM" src="https://github.com/user-attachments/assets/7347d084-40a2-41be-ac9c-7a67001f22a0">


## Why are these changes being made?


*  the selected site masterbar item is not appearing when expected.

## Testing Instructions



* Select a site and go to my home.
* Click on the woo commerce tab and verify the site menu is still present in the masterbar.
* Click back to "all sites", verify the site menu is still present. (same should be true with /read and /me).
* Verify the +new option from here links to editor URLs for the selected site, instead of generic urls that would open the site selector.
* In all sites, select a site in the dashboard. Verify the sidebar collapse animation still works as expected.
* Test around the app and verify the site menu remains in place until the selected site is cleared. Test reloading the app at various places and ensure things look good (reload at /sites, reload at /home before going to sites, etc. - there were some edge cases I resolved where the collapsed sidebar would still appear at unexpected times).

## Pre-merge Checklist
